File: typescript/src/crossover.ts
import { mutation } from "./genome";
import { SelectionItem, Stress, Offspring, Genome } from "./types";

function chunks<T>(table: T[], size: number): T[][] {
  return Array.from({ length: Math.ceil(table.length / size) }).map((_, i) =>
    table.slice(size * i, size * (i + 1))
  );
}

// export function k_points_crossover(
//   A: SelectionItem,
//   B: SelectionItem,
//   stress_A: Stress,
//   stress_B: Stress,
//   k: number = 0
// ): [Offspring, Offspring] {
//   const nbOfPoints = k > 0 ? k : Math.floor(Math.random() * 10)
//   ********To implement*********
// }

export function uniform_crossover(
    A: SelectionItem,
    B: SelectionItem,
    stress_A: Stress,
    stress_B: Stress,
    chunk_size: number = 1
  ): [Offspring, Offspring] {
    const gene_a = chunks(mutation(A[0], stress_A)[1], chunk_size);
    const gene_b = chunks(mutation(B[0], stress_B)[1], chunk_size);
    const crossed = Array.from({ length: gene_a.length }, (_, i) =>
      Math.random() > 0.5 ? [gene_b[i], gene_a[i]] : [gene_a[i], gene_b[i]]
    ).flat(Infinity);
    const new_A = [A[0][0], crossed[0]] as Genome;
    const new_B = [B[0][0], crossed[1]] as Genome;
    return [
      [new_A, 0],
      [new_B, 0],
    ];
  }

export function uniform_independant_crossover(
  A: SelectionItem,
  B: SelectionItem,
  stress_A: Stress,
  stress_B: Stress
): [Offspring, Offspring] {
  const gene_a = chunks(mutation(A[0], stress_A)[1], 20);
  const gene_b = chunks(mutation(B[0], stress_B)[1], 20);
  const new_A = [
    A[0][0],
    Array.from({ length: gene_a.length }, (_, i) =>
      Math.random() > 0.5 ? gene_b[i] : gene_a[i]
    ).flat(Infinity),
  ] as Genome;
  const new_B = [
    B[0][0],
    Array.from({ length: gene_a.length }, (_, i) =>
      Math.random() > 0.5 ? gene_a[i] : gene_b[i]
    ).flat(Infinity),
  ] as Genome;
  return [
    [new_A, 0],
    [new_B, 0],
  ];
}
export function uniform_independant_bias_crossover(
  A: SelectionItem,
  B: SelectionItem,
  stress_A: Stress,
  stress_B: Stress,
  chunk_size: number = 1
): [Offspring, Offspring] {
  const half = stress_A / (stress_A + stress_B);
  const gene_a = chunks(mutation(A[0], stress_A)[1], chunk_size);
  const gene_b = chunks(mutation(B[0], stress_B)[1], chunk_size);
  const new_A = [
    A[0][0],
    Array.from({ length: gene_a.length }, (_, i) =>
      Math.random() < half ? gene_a[i] : gene_b[i]
    ).flat(Infinity),
  ] as Genome;
  const new_B = [
    B[0][0],
    Array.from({ length: gene_a.length }, (_, i) =>
      Math.random() > half ? gene_a[i] : gene_b[i]
    ).flat(Infinity),
  ] as Genome;
  return [
    [new_A, 0],
    [new_B, 0],
  ];
}

File: cpp/crossover.cc
#include <array>
#include <vector>
#include "network.cc"

struct Individual
{
    Genome genome;
    float score;
    float distance;
    Individual() : genome(Genome()), score(0), distance(0){};
    Individual(Genome g) : genome(g), score(0), distance(0){};
    Individual(Genome g, float s, float d) : genome(g), score(s), distance(d){};
    Individual(std::vector<int> t) : genome(randomGenome(t)), score(0), distance(0){};
    Individual(const Individual& i): genome(i.genome), score(i.score), distance(i.distance){};
};

template <typename T>
std::vector<std::vector<T>> chunks(const std::vector<T> table, long long size)
{
    std::vector<T> tmp = table;
    std::vector<std::vector<T>> chunks = {{}};
    while(tmp.size()) {
        if(chunks.back().size() < size){
            chunks.back().push_back(tmp[0]);
        }else {
            chunks.push_back({ tmp[0] });
        }
        tmp.erase(tmp.begin());
    }
    return chunks;
}

std::vector<Individual> uniform_independant_bias_crossover(
    Individual A,
    Individual B,
    float stress_A,
    float stress_B,
    long long chunk_size = 1)
{
    float half = stress_A / (stress_A + stress_B);
    std::vector<std::vector<float>> gene_a = chunks(mutation(A.genome, stress_A).genes, chunk_size);
    std::vector<std::vector<float>> gene_b = chunks(mutation(B.genome, stress_B).genes, chunk_size);
    std::vector<float> new_gene_a;
    std::vector<float> new_gene_b;
    for (long long i = 0; i < gene_a.size(); i++)
    {
        if (((float)rand() / RAND_MAX) < half)
        {
            std::for_each(
                gene_a[i].begin(),
                gene_a[i].end(),
                [&](float g)
                { new_gene_a.push_back(g); });
        }
        else
            std::for_each(
                gene_b[i].begin(),
                gene_b[i].end(),
                [&](float g)
                { new_gene_a.push_back(g); });
        if (((float)rand() / RAND_MAX) < half)
        {
            std::for_each(
                gene_b[i].begin(),
                gene_b[i].end(),
                [&](float g)
                { new_gene_b.push_back(g); });
        }
        else
            std::for_each(
                gene_a[i].begin(),
                gene_a[i].end(),
                [&](float g)
                { new_gene_b.push_back(g); });
    }
    return {
        Individual(Genome(A.genome.topology, new_gene_a)),
        Individual(Genome(B.genome.topology, new_gene_b))};
}
